src/main/java/com/pingcap/tikv/operation/IndexScanIterator.java

```diff
@@ -20,38 +20,89 @@
 import com.pingcap.tikv.Snapshot;
 import com.pingcap.tikv.codec.KeyUtils;
 import com.pingcap.tikv.codec.TableCodec;
+import com.pingcap.tikv.grpc.Coprocessor;
 import com.pingcap.tikv.grpc.Coprocessor.KeyRange;
 import com.pingcap.tikv.meta.TiSelectRequest;
 import com.pingcap.tikv.row.Row;
-import java.util.Iterator;
+import com.pingcap.tikv.util.Comparables;
+
+import java.util.*;
+import java.util.stream.Collectors;
 
 // A very bad implementation of Index Scanner barely made work
 // TODO: need to make it parallel and group indexes
 public class IndexScanIterator implements Iterator<Row> {
   private final Iterator<Row> iter;
   private final TiSelectRequest selReq;
   private final Snapshot snapshot;
+  private final boolean singleRead;
 
-  public IndexScanIterator(Snapshot snapshot, TiSelectRequest req, Iterator<Row> iter) {
+  public IndexScanIterator(
+      Snapshot snapshot, TiSelectRequest req, Iterator<Row> iter, boolean singleRead) {
     this.iter = iter;
     this.selReq = req;
     this.snapshot = snapshot;
+    this.singleRead = singleRead;
   }
 
   @Override
   public boolean hasNext() {
     return iter.hasNext();
   }
 
+  private List<KeyRange> mergeKeyRangeList(List<KeyRange> keyRangeList) {
+    // sort key ranges according to its start key in order to do further merge.
+    List<KeyRange> newKeyRanges = new LinkedList<>();
+    keyRangeList.sort((a, b) -> Comparables.wrap(a.getStart()).compareTo(b.getStart()));
```

Comment:
here you might wrap all together to sorted first otherwise each compare it creates an object.